### ISSUE TITLE
Confused by network behaviour

### ISSUE THREAD

**Issue by ashic** [HubSpot/Singularity#590]
Hi,
I'm trying out singularity, and so far it looks good. I'm trying to sort out networking, but it's baffling me:

* I've created a single instance service that's not load balanced. In deploy, I specify docker: tutum/hello-world , bridge network with the following:

```
Container port type: Literal, 80
Host port type: Literal, 31010
```

When I do this, I get a response back from hostname:80, but not from hostname:31010. I should add that --resources=ports:[8000-8050,31000-32000] is specified when launching the slave.

* I'd really like to be able to hit a service endpoint instead of slave machines. I see that if I select "load balanced" when creating the request, I get a service base path (e.g. /singularity/api/v2) option when deploying. However, I've not been able to get this to work. Do I need something else set up for this to work? I don't have mesos-dns setup. Will that help? Will https://github.com/CiscoCloud/mesos-consul be useful for this as well?

* When selecting Bridge Network and adding a Docker Port Mapping, I see that using 
```
Container Port Type: Literal, 80
Host Port Type: From Offer
Container Port: 0
Protocol: tcp
```
gives me the following error:
```
There was a problem: Index of port resource for hostPort must be between 0 and -1 (inclusive)
```

Is this expected? Is there some documentation as to how docker networking relates to Singularity?

Thanks,
Ashic.

**Comment by ashic** [comment]
Just solved the third bit: needed to ask for 1 port to be able to do From Offer[0]. First two points are still confusing me though. Tried the following:

```
"containerInfo": {
            "type": "DOCKER",
            "docker": {
                "image": "tutum/hello-world",
                "privileged": false,
                "network": "BRIDGE",
                "portMappings": [
                    {
                        "containerPortType": "LITERAL",
                        "containerPort": 80,
                        "hostPortType": "LITERAL",
                        "hostPort": 8050,
                        "protocol": "tcp"
                    }
                ]
            }
        },
```

Yet still, curl localhost:8050 fails, while curl localhost:80 succeeds. Any ideas?

**Comment by ssalinas** [comment]
To speak to your first point, there are a number of options for load balancing services run in mesos. We use Singularity with [Baragon](https://github.com/HubSpot/Baragon) and nginx, which allows each load balanced service running in mesos to be available at a certain path and Baragon keeps the host:port of the running services in sync with Singularity. Docs for that can be found [here](https://github.com/HubSpot/Baragon/blob/master/docs/managing_baragon.md#singularity) and [here](https://github.com/HubSpot/Singularity/blob/cc84580434f85d6dff63d3f771ff3d42a08b2188/Docs/development/lbs.md)

I am still investigating your second question. It seems that you have it correct and I will let you know if I find anything

**Comment by stevenschlansker** [comment]
I'm not at all sure this is related, but `BRIDGE` mode seems to not be effective for us:

```
{
  "healthcheckUri": null,
  "requestId": "teamcity-agent",
  "resources": {
    "cpus": 1,
    "memoryMb": 4096,
    "numPorts": 1
  },
  "containerInfo": {
    "docker": {
      "privileged": true,
      "network": "BRIDGE"
    },
    "volumes": [
      {
        "hostPath": "/opt/ot",
        "containerPath": "/opt/ot",
        "mode": "RO"
      }
    ]
  }
}
```

creates a container as such:

```
root@mesos-slave4-qa-uswest2:~# docker inspect 2d5f5604e4bb
[
{
    "Id": "2d5f5604e4bb5c4da4ca83538237834479c99248df4e0c2dc8a28ed2ca838e26",
    "Created": "2015-09-04T18:16:38.50261588Z",
    "Path": "/bin/sh",
    "Args": [
        "-c",
        "/var/teamcity/build-agent/bin/buildAgentWrapper.sh"
    ],
    "State": {
        "Running": true,
        "Paused": false,
        "Restarting": false,
        "OOMKilled": false,
        "Dead": false,
        "Pid": 25477,
        "ExitCode": 0,
        "Error": "",
        "StartedAt": "2015-09-04T18:16:38.575677926Z",
        "FinishedAt": "0001-01-01T00:00:00Z"
    },
    "Image": "4bf6034147655e0f9cd37ccf795c7c6beb21f6992890fc640387eb3d497c3e2d",
    "NetworkSettings": {
        "Bridge": "",
        "EndpointID": "4186dff0c063a42f3bb46e9b423b417bab2466472c640f3a0bbedb219ef149a9",
        "Gateway": "\u003cnil\u003e",
        "GlobalIPv6Address": "",
        "GlobalIPv6PrefixLen": 0,
        "HairpinMode": false,
        "IPAddress": "",
        "IPPrefixLen": 0,
        "IPv6Gateway": "",
        "LinkLocalIPv6Address": "",
        "LinkLocalIPv6PrefixLen": 0,
        "MacAddress": "",
        "NetworkID": "02178ae9fb3961cb5bd659dcad35cf304a850a0626eb5453ba7e4df2c8c9608c",
        "PortMapping": null,
        "Ports": {},
        "SandboxKey": "/var/run/docker/netns/default",
        "SecondaryIPAddresses": null,
        "SecondaryIPv6Addresses": null
    },
    "Name": "/mesos-79aa5220-c69b-4262-8e13-ace5d98cc227",
        "NetworkMode": "host",
```

Again, we haven't proven that this is the problem -- but @ashic could you provide a `docker inspect` output for one of the containers with this problem?

**Comment by ssalinas** [comment]
Issues addressed here should be fixed by #837
